Improve MSW menu dark-mode rendering and Windows 11 menu appearance

memoarfaa opened the pull request:
## Description
This PR fixes several MSW menu rendering issues observed in dark mode and on Windows 11. 
1- It includes changes to make menu breaks render correctly in dark mode (owner-drawn).
2- applies dark-mode-aware menu background colors, adjusts owner-drawn menu item drawing (text color fallbacks).
3- applies Windows 11 DWM window attributes to popup menu windows to get rounded corners and proper border color.

High-level items implemented or touched by this PR:
- Force owner-draw for menu-break items when dark mode is active so they render correctly.
- Use theme-aware colors for popup menu background and menu item text; add fallbacks when dark-theme values are unavailable.
- change popup menu background .
- Update theme handle usage (supply dark-mode theme identifiers where applicable).
- Set DWM attributes (rounded corners and border color) for menu windows on Windows 11+ dynamically.

## Changes Proposed
- src/msw/menu.cpp
  - Include darkmode and uxtheme private headers.
  - When m_doBreak is set and dark mode is active, mark the item owner-drawn.
  - On menu creation, when wxMSWDarkMode is active and owner-drawn is used, query theme popup background color.
- src/msw/menuitem.cpp
  - Include darkmode header.
  - Use dark-mode-aware theme string when creating wxUxThemeHandle).
  - Obtain TMT_TEXTCOLOR from theme in dark mode, with fallback to system menu text color if not available.
- src/msw/window.cpp
  - Dynamically call DwmSetWindowAttribute to set DWMWA_WINDOW_CORNER_PREFERENCE (rounded corners) and DWMWA_BORDER_COLOR.
  - Load dwmapi.dll and DwmSetWindowAttribute with LoadLibrary/GetProcAddress to avoid link dependency.

## Motivation and Context
- Dark-mode menus and menu breaks displayed incorrectly with default rendering; forcing owner-draw and using theme colors addresses incorrect backgrounds and text colors in dark themes.
- Windows 11 introduced different menu visuals (rounded corners, border colors). Applying DWM attributes to popup menu windows improves visual consistency for native-looking menus.
- The changes aim to preserve visual correctness across themes and OS versions while avoiding hard dependencies on newer APIs at link time.

## How Has This Been Tested?
- [x] Built and ran on Windows 11 (verify rounded corners and border color on popup menus).
- [x] Built and ran on Windows 10 (verify no regressions; attributes not applied).
- [x] Tested owner-drawn menu breaks in dark mode and light mode (verify layout and separators).
- [x] Verified submenu arrows render for owner-drawn popup items at multiple DPIs.
- [x] Verified no GDI resource leaks (monitor GDI handles while repeatedly opening/closing menus).

## Screenshots
before
<img width="707" height="530" alt="Screenshot 2025-12-24 101158" src="https://github.com/user-attachments/assets/9c3cc313-1b6f-4497-8524-ea1ca662efce" />

after 
<img width="698" height="606" alt="Screenshot 2025-12-24 100600" src="https://github.com/user-attachments/assets/470304e1-1531-48e3-8239-6d41675fb6e5" />


## Developer notes and required follow-ups
Address the following soon after merging this PR:


- Initialization placement
  - The menu background setting is done in Create(); evaluate whether it should be moved to wxMenu::Init() or a more central 


## Related issues
Fixes #25577
Fixes #22518

## Suggested reviewers
@vadz , @PBfordev

memoarfaa committed:
Fix menu rounded corners in Windows 11 and higher

memoarfaa committed:
Set dark mode menu background color and if menu  has break Set it OwnerDrawn at darkMode

memoarfaa committed:
Add DarkMode Theme to OwnerDrawn Menu Items

memoarfaa committed:
fix trailing whitespace build errors

memoarfaa committed:
add submenu arrows render for owner-drawn popup items at dark mode.

memoarfaa committed:
Fix build errors

memoarfaa committed:
Fix IsActive missing "()" build error

PBfordev commented on src/msw/menu.cpp:
```diff
@@ -916,30 +922,47 @@ void wxMenuBar::Refresh()
 WXHMENU wxMenuBar::Create()
 {
-    if ( m_hMenu != 0 )
+    if (m_hMenu != 0)
```
There is a bunch of whitespace changes going against the code style (perhaps done automatically by a tool). These should be reverted (and the new code formatting should match the expected style).

PBfordev commented on src/msw/menu.cpp:
```diff
@@ -916,30 +922,47 @@ void wxMenuBar::Refresh()
+            COLORREF crMenu = wxColourToRGB(hTheme.GetColour(MENU_POPUPBACKGROUND, TMT_FILLCOLOR));
+            WinStruct<MENUINFO> mi;
+            mi.fMask = MIM_BACKGROUND | MIM_APPLYTOSUBMENUS;
+            // Maybe we should clean up this brush to avoid GDI leaks?
```
Perhaps we could use `wxTheBrushList->FindOrCreateBrush()` here?

PBfordev commented on src/msw/menu.cpp:
```diff
@@ -916,30 +922,47 @@ void wxMenuBar::Refresh()
+            // Set dark mode menu background color
+            // May be this should be done in wxMenu::Init() instead?
+            wxUxThemeHandle hTheme(GetFrame()->AsWindow(), L"MENU", L"DarkMode_ImmersiveStart::Menu");
+            COLORREF crMenu = wxColourToRGB(hTheme.GetColour(MENU_POPUPBACKGROUND, TMT_FILLCOLOR));
```
I am not sure about getting the color from the theme (here and elsewhere): How does `wxDarkModeSettings::GetColour()` is expected to be used?

PBfordev commented on src/msw/window.cpp:
```diff
@@ -3598,7 +3599,63 @@ wxWindowMSW::MSWHandleMessage(WXLRESULT *result,
         case WM_ENDSESSION:
             processed = HandleEndSession(wParam != 0, lParam);
             break;
+        case WM_ENTERIDLE:
```
I have no idea if doing this here is where it should be done.

But I wonder if we should do this:
1. For every idle message?
2. When not in dark mode?
3. If not Windows 11?

vadz replied:
I agree with the points (2) and (3), i.e. we need to restrict this to dark mode under Windows 11 (and later). Point (1) is not relevant, I think, as we don't do this on every idle message, but only when entering a local message loop.

OTOH I'd like to add another point: it would be nice to move this into a separate function.

PBfordev commented on src/msw/window.cpp:
```diff
@@ -3598,7 +3599,63 @@ wxWindowMSW::MSWHandleMessage(WXLRESULT *result,
+                                    //    SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER  | SWP_FRAMECHANGED);
+                                }
 
+                                // Apply border color to menu window
```
I have noticed that now (compared to the master), all the menus (including system window menu) have very light border. No such border is used by menus in other applications shipped with Windows.

PBfordev reviewed (commented):
Thanks for the work. 

I am not really qualified to review this, so my comments are just nitpicks and questions.

vadz commented on src/msw/menuitem.cpp:
```diff
@@ -972,6 +973,19 @@ bool wxMenuItem::OnDrawItem(wxDC& dc, const wxRect& rc,
             }
 
             hTheme.DrawBackground(hdc, rcSelection, MENU_POPUPITEM, state);
+            // we need also to draw menu arrow if the menu item at popup menu and has subMenu for dark mode.
```
If I understand the meaning of this comment correctly, it should say
```suggestion
            // we need also to draw menu arrow of sub-menus in dark mode
```

vadz reviewed (commented):
Thanks a lot, this is definitely very useful and should be merged. It would be great if you could please address the comments, but if not I'll try doing it myself later.

memoarfaa commented:
> @memoarfaa Do you plan to make more changes here or should I take it over?
>
Yes give me some time

memoarfaa committed:
Apply suggestions from code review

The "#include" error that was created using suggestions from Visual Studio has been corrected.

Co-authored-by: VZ <vz-github@zeitlins.org>

memoarfaa committed:
Add Light/DarkTheme for menuitem

memoarfaa committed:
Add Light/Dark Theme to menu background color

memoarfaa committed:
Add HandleEnterIdle for rounded corners in Windows 11 and higher at light mode if menu has Owner Drawn items.

memoarfaa committed:
Add  WM_MENUBAR_INITMENU Message For Enable Round Corners For Menu at DarkMode

memoarfaa committed:
Add Header Items.

memoarfaa committed:
resolve conflicts

memoarfaa committed:
Fix build errors

memoarfaa committed:
Fix build errors

memoarfaa committed:
Fix trailing whitespace

memoarfaa committed:
fix Window.cpp trailing whitespace.

memoarfaa committed:
Fix EnableRoundCorners error

memoarfaa committed:
Merge branch 'wxWidgets:master' into darkmode

memoarfaa committed:
Add "brush.h"

memoarfaa committed:
improve hWndMenu

memoarfaa committed:
Merge branch 'darkmode' of https://github.com/memoarfaa/wxWidgets into darkmode
